File: backend/src/main/java/com/alexandria/service/AvaliacaoService.java
package com.alexandria.service;

import com.alexandria.dto.avaliacao.AtualizarAvaliacaoRequest;
import com.alexandria.dto.avaliacao.AvaliacaoResponse;
import com.alexandria.dto.avaliacao.CriarAvaliacaoRequest;
import com.alexandria.dto.livro.LivroResumoResponse;
import com.alexandria.exception.DuplicateResourceException;
import com.alexandria.exception.ResourceNotFoundException;
import com.alexandria.model.Avaliacao;
import com.alexandria.model.Livro;
import com.alexandria.model.User;
import com.alexandria.repository.AvaliacaoRepository;
import com.alexandria.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class AvaliacaoService {

    private final AvaliacaoRepository avaliacaoRepository;
    private final UserRepository userRepository;
    private final LivroService livroService;
    private final GamificacaoService gamificacaoService;
    private final ComunidadeService comunidadeService;
    private final CsvExportService csvExportService;

    public AvaliacaoService(AvaliacaoRepository avaliacaoRepository,
                             UserRepository userRepository,
                             LivroService livroService,
                             GamificacaoService gamificacaoService,
                             ComunidadeService comunidadeService,
                             CsvExportService csvExportService) {
        this.avaliacaoRepository = avaliacaoRepository;
        this.userRepository = userRepository;
        this.livroService = livroService;
        this.gamificacaoService = gamificacaoService;
        this.comunidadeService = comunidadeService;
        this.csvExportService = csvExportService;
    }

    @Transactional
    public AvaliacaoResponse criar(Long usuarioId, CriarAvaliacaoRequest request) {
        Livro livro = livroService.obterOuCriarLivroLocal(request.identificadorExterno());

        if (avaliacaoRepository.existsByUsuarioIdAndLivroId(usuarioId, livro.getId())) {
            throw new DuplicateResourceException("Você já avaliou este livro.");
        }

        User usuario = userRepository.findById(usuarioId)
                .orElseThrow(() -> new ResourceNotFoundException("Usuário não encontrado."));

        Avaliacao avaliacao = Avaliacao.builder()
                .usuario(usuario)
                .livro(livro)
                .nota(request.nota())
                .resenha(request.resenha())
                .build();
        avaliacao = avaliacaoRepository.save(avaliacao);

        gamificacaoService.avaliacaoCriada(usuarioId);
        comunidadeService.criarPostAutomatico(usuario, livro, avaliacao);

        return paraResponse(avaliacao);
    }

    public List<AvaliacaoResponse> listar(Long usuarioId) {
        return avaliacaoRepository.findByUsuarioIdOrderByCriadoEmDesc(usuarioId).stream()
                .map(this::paraResponse)
                .toList();
    }

    @Transactional
    public AvaliacaoResponse atualizar(Long usuarioId, Long avaliacaoId, AtualizarAvaliacaoRequest request) {
        Avaliacao avaliacao = buscarAvaliacao(usuarioId, avaliacaoId);
        avaliacao.setNota(request.nota());
        avaliacao.setResenha(request.resenha());
        return paraResponse(avaliacao);
    }

    @Transactional
    public void remover(Long usuarioId, Long avaliacaoId) {
        Avaliacao avaliacao = buscarAvaliacao(usuarioId, avaliacaoId);
        avaliacaoRepository.delete(avaliacao);
        gamificacaoService.avaliacaoRemovida(usuarioId);
    }

    public String exportarCsv(Long usuarioId) {
        List<Avaliacao> avaliacoes = avaliacaoRepository.findByUsuarioIdOrderByCriadoEmDesc(usuarioId);
        return csvExportService.exportarAvaliacoes(avaliacoes);
    }

    private Avaliacao buscarAvaliacao(Long usuarioId, Long avaliacaoId) {
        return avaliacaoRepository.findByIdAndUsuarioId(avaliacaoId, usuarioId)
                .orElseThrow(() -> new ResourceNotFoundException("Avaliação não encontrada."));
    }

    private AvaliacaoResponse paraResponse(Avaliacao avaliacao) {
        Livro livro = avaliacao.getLivro();
        LivroResumoResponse livroResumo = new LivroResumoResponse(
                livro.getIdentificadorExterno(),
                livro.getTitulo(),
                livro.getAutor(),
                livro.getCapa(),
                livro.getCategoria(),
                livro.getDataPublicacao()
        );
        return new AvaliacaoResponse(
                avaliacao.getId(),
                livroResumo,
                avaliacao.getNota(),
                avaliacao.getResenha(),
                avaliacao.getCriadoEm(),
                avaliacao.getAtualizadoEm()
        );
    }
}

File: backend/src/services/avaliacaoService.js
const prisma = require('../lib/prisma')
const livroService = require('./livroService')
const gamificacaoService = require('./gamificacaoService')
const comunidadeService = require('./comunidadeService')
const csvExportService = require('./csvExportService')
const { DuplicateError, NotFoundError } = require('../lib/errors')

function paraResponse(avaliacao) {
  return {
    id: avaliacao.id,
    livro: {
      identificadorExterno: avaliacao.livro.identificadorExterno,
      titulo: avaliacao.livro.titulo,
      autor: avaliacao.livro.autor,
      capa: avaliacao.livro.capa,
      categoria: avaliacao.livro.categoria,
      dataPublicacao: avaliacao.livro.dataPublicacao,
    },
    nota: avaliacao.nota,
    resenha: avaliacao.resenha,
    criadoEm: avaliacao.criadoEm,
    atualizadoEm: avaliacao.atualizadoEm,
  }
}

async function criar(usuarioId, { identificadorExterno, nota, resenha }) {
  const livro = await livroService.obterOuCriarLivroLocal(identificadorExterno)

  const existente = await prisma.avaliacao.findUnique({
    where: { usuarioId_livroId: { usuarioId, livroId: livro.id } },
  })
  if (existente) {
    throw new DuplicateError('Você já avaliou este livro.')
  }

  const usuario = await prisma.user.findUnique({ where: { id: usuarioId } })
  if (!usuario) {
    throw new NotFoundError('Usuário não encontrado.')
  }

  const avaliacao = await prisma.avaliacao.create({
    data: { usuarioId, livroId: livro.id, nota, resenha },
    include: { livro: true },
  })

  await gamificacaoService.avaliacaoCriada(usuarioId)
  await comunidadeService.criarPostAutomatico(usuario, livro, avaliacao)

  return paraResponse(avaliacao)
}

async function listar(usuarioId) {
  const avaliacoes = await prisma.avaliacao.findMany({
    where: { usuarioId },
    include: { livro: true },
    orderBy: { criadoEm: 'desc' },
  })
  return avaliacoes.map(paraResponse)
}

async function atualizar(usuarioId, avaliacaoId, { nota, resenha }) {
  await buscarAvaliacao(usuarioId, avaliacaoId)

  const atualizada = await prisma.avaliacao.update({
    where: { id: avaliacaoId },
    data: { nota, resenha },
    include: { livro: true },
  })

  return paraResponse(atualizada)
}

async function remover(usuarioId, avaliacaoId) {
  await buscarAvaliacao(usuarioId, avaliacaoId)
  await prisma.avaliacao.delete({ where: { id: avaliacaoId } })
  await gamificacaoService.avaliacaoRemovida(usuarioId)
}

async function exportarCsv(usuarioId) {
  const avaliacoes = await prisma.avaliacao.findMany({
    where: { usuarioId },
    include: { livro: true },
    orderBy: { criadoEm: 'desc' },
  })
  return csvExportService.exportarAvaliacoes(avaliacoes)
}

async function buscarAvaliacao(usuarioId, avaliacaoId) {
  const avaliacao = await prisma.avaliacao.findFirst({ where: { id: avaliacaoId, usuarioId } })
  if (!avaliacao) {
    throw new NotFoundError('Avaliação não encontrada.')
  }
  return avaliacao
}

module.exports = { criar, listar, atualizar, remover, exportarCsv }
